### ISSUE TITLE
Peer command lets user specify options that don't make sense without any warnings

### ISSUE THREAD

**Issue by lehors** [hyperledger-archives/fabric#1309]
Privacy/confidentiality only works when security is enabled. When privacy is on and security is off the peer command silently ignores the privacy setting.
Similarly when security is off, if a username is specified it is silently ignored.
It would be more user friendly to let the user know about those situations when they occur rather than ignoring them silently so that they can be corrected if necessary.

Alternatively, security could be automatically turned on when privacy is requested or a username is specified.

**Comment by JonathanLevi** [comment]
+1

Yes, the 'mode' should probably be more explicit and sometimes even warn the users about some (less recommended) combinations.

**Comment by JonathanLevi** [comment]
Along these lines: I have noticed something similar relating to Client Authentication and thought the same. I will open a separate issue in that regard. While related in nature, that is, being more transparent and clear about what's happening 'behind the scene', yet still separate enough to justify its own issue/discussion/fix.

**Comment by lehors** [comment]
Hi @JonathanLevi, thanks for your feedback. Do you have a preference over which way to fix this? I have a patch to just spit out warnings which I think is an improvement already but I wonder about the alternative. The thing is that it can be convenient to just turn everything off by simply turning security off. In that case, turning it on automatically wouldn't be seen as an improvement.

**Comment by JonathanLevi** [comment]
Yes, I agree: first, a warning is already an improvement!

As a user, I would rather know what's going on (so I'd rather take the conservative/restrictive approach). 

Effectively, we can think about privacy/confidentiality as a 'sub-item' or 'sub-category' of security, that is:
1. Security (On/Off)
   1.1 Privacy (On/Off)

Please try to picture it with some fancy Javascript/dynamic check-boxes that make the sub-item 1.1 Privacy disabled when Security is unchecked. Please pardon my (lack of) visual animation/design skills.

Where we do **not allow** at all for privacy to be ON when security is OFF, and simply break the flow.

Would you agree?

**Comment by lehors** [comment]
By "not allow" do you mean it should generate an error and stop, rather than producing a warning and continuing?

Because it is pretty much the case currently that privacy is not allowed to be ON when security is OFF. The problem is that it is done by silently ignoring the privacy setting.

Here is what my patch does (with security and privacy off by default):

> $ CORE_SECURITY_PRIVACY=true ./peer peer --peer-chaincodedev
> ....
> 01:13:46.185 [main] serve -> INFO 00a Security enabled status: false
> 01:13:46.186 [main] serve -> INFO 00b Privacy enabled status: true
> 01:13:46.186 [main] serve -> WARN 00c For privacy to be enabled security must be enabled.
> ...
> 
> $ CORE_SECURITY_PRIVACY=true ./peer chaincode deploy -u jim -n mycc -c '{"Function":"init", "Args": ["a","100", "b", "200"]}'
> ...
> 01:14:30.831 [crypto] main -> INFO 001 Log level recognized 'info', set to INFO
> 01:14:30.845 [main] chaincodeDeploy -> WARN 002 Username supplied but security is disabled.
> 01:14:30.845 [main] chaincodeDeploy -> WARN 003 For privacy to be enabled security must be enabled.
> ...

**Comment by JonathanLevi** [comment]
Yes, by break the flow I mean to stop/exit/panic. Especially as it states that 'For privacy to be enabled security **must** be enabled' and it was not.

A warning is not enough IMHO (even/regardless of the log verbosity level).

**Comment by lehors** [comment]
Understood. @binhn, @jeffgarratt, @srderson what do you think about that?

**Comment by Nyaosi** [comment]
if (security enabled status is false) then (privacy enabled status 
should be false too), it's not accurate to display it as true

**Comment by corecode** [comment]
Additionally, misspelled keys in the config are ignored.

**Comment by lehors** [comment]
@corecode would a warning a la "unrecognized configuration: %%%" work for you or do you think it should fail with an error in this case?

We may not want to fail on this, to be more future proof. Say we add some setting in version x that isn't known by version x-1, may be we still want to be able to run version x-1 with the new config?

**Comment by corecode** [comment]
I prefer failing - better than half working (imagine running this from a script - you may never see the warnings).

**Comment by lehors** [comment]
I looked into this a bit more and I can easily add a few tests to catch some situations that are wrong and let the user know about those. Looking for misspelled keys and values on the other hand isn't cheap.

First, this requires having a canonical list of keys and values that can be used a reference. Then, it requires checking that all the keys and values provided exist in that list.

Configuration is currently handled with viper which for instance returns any boolean values that don't match "true" as false. So, if you have "tue" for instance in your config file, viper will simply report it as false. Catching this requires getting the string value instead of the boolean and checking that is indeed either "true" or "false" and report as an error anything else.

Similarly, keys are queried by the application as in viper.GetBool("security.enabled"). If you mistyped that key in your config file and put, say, "security.enalbed", that key is ignored and security.enabled is simply reported as false. Catching this requires walking through all the keys read in viper and checking that they exist in the canonical list.

There is nothing complicated here but it does require 1) maintaining a reference list and 2) doing some additional checking. I'd like to hear from others whether this is considered worthwhile.

One possibility would be to do these extra checks only when in some special mode, like debugging, - it would require maintaining the canonical list though - or to do that with a different program altogether. Like a lint for fabric's configuration files if you will.

Again, I'd like input from others on this one.
Thanks.

**Comment by corecode** [comment]
Or switch from viper to something else, more strict.

**Comment by lehors** [comment]
Indeed, that's another possibility.

**Comment by srderson** [comment]
If privacy requires security, perhaps it should be one setting with three options?

In terms of catching bad input, I'd agree with @corecode about switching to something more strict. Maybe there's some sort of Go YAML validation library that could be run at startup?

**Comment by JonathanLevi** [comment]
@lehors Yes, I like the ability to run something like 'lint', or any (optional) check with verbose output.
It's a bit scary indeed that "Tue" suddenly means "Yes" ;-)

@corecode, @srderson I, too, second a more strict approach (along the lines of the early suggestion to throw an exception rather than 'silently' warn...). I look at (YAML) conf files as a user-define input, which certainly requires proper 'input validation' (as a step/library or otherwise).

Seems like there are many such case. We should list out (@lehors' reference list) and tackle those set of configurations that 'do not make sense' as per the title of this. It might be an on-going loop of:
1. Identify configurations that don't make sense
2. Describe and discuss
3. Break to actionable issues/tasks
4. Fix (and add the logic to @srderson 's suggested YAML validation component + document)
5. Test (positive flow + negative flow) [Should be part of 5, but worth highlighting]
6. Repeat

Please let me know if I can assist, especially regarding the security/privacy/crypto side.

**Comment by corecode** [comment]
Relevant SO post: http://stackoverflow.com/questions/16465705/how-to-handle-configuration-in-go

I think we should not limit ourselves to yaml if other packages provide obvious benefits.  Transitioning config file structure is painful, but only once.  Generally, people on go-nuts seem to dislike yaml, because it is a very complex format.

Options I can see:
- https://github.com/BurntSushi/toml ini-like, uses reflection, can express complex data (arrays, time)
- https://godoc.org/gopkg.in/gcfg.v1 ini, uses reflection, can express multi-section and multi-values
- https://golang.org/pkg/encoding/json/ json, standard go package, uses reflection, complex data, no comments
- https://github.com/vharitonsky/iniflags ini, ties into standard go flags package, allows overrides via command line

**Comment by christo4ferris** [comment]
from the go-nuts thread; https://groups.google.com/d/msg/golang-nuts/sAofmg2_lL8/_7ctHwssGxwJ "YAML is the Perl of data formats" 

ROTFLMAO

**Comment by bcbrock** [comment]
I second ( #1962 ) third ( #1205 ) and fourth ( #647 ) the idea to ditch Viper.

I like the idea of having a file that lists all of the valid keys with their default values, as this provides an opportunity to document them all. The current YAML files could serve this purpose, however users are encouraged to modify these files, which potentially requires doing a little dance every time a user's copy of the codebase is updated. I would like to see something like a "read-only", sacrosanct, blessed-by-the-elders file of defaults, with a mechanism for a user to create a set of private overrides in an independent file.

**Comment by christo4ferris** [comment]
Let's use this one issue to track everyone's antipathy towards viper and the desire to possibly refactor and significantly improve the whole configuration situation (which BTW is completely broken from my perspective from an operations consideration). related but now closed issues include #1561, #1962 #1205 #647 (thanks @bcbrock ) and probably others.

**Comment by christo4ferris** [comment]
See #1588 as we need to address that issue in any refactor.

**Comment by lehors** [comment]
This issue has been transferred to Jira:
https://jira.hyperledger.org/browse/FAB-50
